src/value/dumper.rs

```diff
@@ -0,0 +1,505 @@
+                        af_writer.write_string(try_to!(element.as_str()))?;
+                    }
+                    IonType::Clob => {
+                        af_writer.write_clob(try_to!(element.as_bytes()))?;
+                    }
+                    IonType::Blob => {
+                        af_writer.write_blob(try_to!(element.as_bytes()))?;
+                    }
+                    IonType::List | IonType::SExpression => {
+                        af_writer.start_container(ion_type.into())?;
+                        {
+                            let seq = try_to!(element.as_sequence());
+                            for child in seq.iter() {
+                                self.write(child)?;
+                            }
+                        }
+                        self.writer.finish_container()?;
+                    }
+                    IonType::Struct => {
+                        af_writer.start_container(ion_type.into())?;
+                        {
+                            let structure = try_to!(element.as_struct());
+                            for (field_name_token, child) in structure.iter() {
+                                let field_name = try_to!(field_name_token.text());
+                                self.write_element(Some(field_name), child)?;
+                            }
+                        }
+                        self.writer.finish_container()?;
+                    }
+                }
+            }
+            Ok(())
+        } else {
+            illegal_operation(format!(
+                "Could not serialize annotation(s) with no text: {:?}",
+                element
+            ))
+        }
+    }
+}
+
+impl<'a> Dumper for IonCWriterFixedDumper<'a> {
+    type Output = &'a [u8];
+
+    #[inline]
+    fn write<E: Element>(&mut self, element: &E) -> IonResult<()> {
+        self.write_element(None, element)
+    }
+
+    fn finish(self) -> IonResult<Self::Output> {
+        if let Some(error) = self.error {
+            return Err(error);
+        }
+
+        // close out the writer and get the written length
+        let data = self.data;
+        let len = {
+            // consume the writer so that we drop it at the end
+            let mut writer = self.writer;
+            writer.finish()?
+        };
+
+        // at this point we can make a slice reference bound to the lifetime parameter
+        // because the writer is no longer borrowing it implicitly (and has been dropped)
+        let output = unsafe { std::slice::from_raw_parts(data, len) };
+        Ok(output)
+    }
+}
+
+/// Whether or not the text is pretty printed or serialized in a more compact representation.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum TextKind {
+    Compact,
+    Pretty,
+}
+
+/// Basic configuration options for [`Dumper`] instances.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum Format {
+    Text(TextKind),
+    Binary,
+    // TODO a mode for Json(TextKind)
```

Comment:
Should the other options be named IonText, IonBinary?

Reply:
I was kind of thinking the _Ion_ part was implied, but am open to it.